Change preprocess API to accept BackendDebugHandleGenerator functor to generates debug handles for module being lowered

kimishpatel opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#58873 [Pytorch backend delegation] Preprocess to accept**

BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

kimishpatel committed:
[Pytorch backend delegation] Preprocess to accept
BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28648613).

kimishpatel committed:
Update on "[Pytorch backend delegation] Preprocess to accept"


BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28648613).

iseeyuan commented on torch/csrc/jit/backends/backend_detail.h:
```diff
@@ -11,7 +12,7 @@ namespace jit {
 
 using BackendPreprocessFunction =
-    std::function<c10::IValue(const Module&, const c10::Dict<IValue, IValue>&)>;
+    std::function<c10::IValue(const Module&, const c10::Dict<IValue, IValue>&, BackendDebugInfoRecorder&)>;
```
Since there's already an argument of `c10::Dict<IValue, IValue>&`, why not we wrap the BackendDebugInfoRecorder into an IValue (which can be anything), and put it into the dict, like `{"BackendDebugInfoRecorder": recorder_obj}`? 

By doing this we don't need to change the signature of `BackendPreprocessFunction`. I think it makes sense to put it into the dict as well, as a "method_compile_spec", or "method_compile_util".

kimishpatel replied:
As far as I understand you cannot wrap any arbitrary cpp object in IValue. For arbitrary cpp objects you have to create custom class out of it first. But if it is possible that would be certainly better.

raziel replied:
Yeah I think that's right, plus it's nice to have it clearly tagged as something that is core to the preprocessing step.

My suggestion though is to not pass the BackendDebugInfoRecorder directly, but have a simple interface that only focus is to define the ```generate_debug_handles()```.

So, more concretely is to have something along the lines of:

```
class DebugHandleManager { // better names appreciated
  virtual NodeToDebugHandle generate_debug_handles(const std::shared_ptr<Graph>& graph);
}

class BackendDebugInfoRecorder {
  ...
  DebugHandleManager get_debug_handle_manager();
}
```

Basically an interface that is isolate and only focuses on generating the debug handles, so the BackendPreprocessFunction is cleaner.

kimishpatel replied:
So I understand the question about cleaner interface that does simply debug handle generation. But I would prefer to not introduce yet another class that does just that. Maybe we can just have:
```
BackendDebugInfoRecorder debug_info_recorder;
auto generate_debug_handles_fn = [&debug_info_recorder](const std::shared_ptr<Graph>& g) {
  return debug_info_recorder.generate_debug_handles(g);
  // or 
 return generate_debug_handles(debug_info_recorder, g);
};
loweredModule.register_attribute("__processed_module", AnyType::get(),    
  detail::getBackendPreprocessFunction(backend_name)( cloned_module, method_compile_spec, 
    generate_debug_handles_fn), 
  /*is_param=*/false);
```

What do you think?

kimishpatel replied:
Yup, that sound good.

raziel commented on torch/csrc/jit/backends/backend_debug_handler.h:
```diff
@@ -125,31 +126,17 @@ using BackendDebugInfoMapType =
   // can throw, and throwing with dtor will call terminate and thus voids any
   // exception catching at a higher level.
   BackendDebugInfoMapType stopRecording();
+  NodeToDebugHandle generate_debug_handles(const std::shared_ptr<Graph>& graph);
```
See my comment below (in the preprocess function) about having a separate object to handle this which the recorder can produce and pass to preprocess functions.

raziel commented on torch/csrc/jit/backends/generate_debug_handles.h:
```diff
@@ -8,8 +8,6 @@
 
-using NodeToDebugHandle = std::unordered_map<Node*, DebugHandleType>;
-
 /*
```
why aren't these files (.h & .cpp) deleted?

kimishpatel replied:
My bad!

raziel commented on test/cpp/jit/test_backend_compiler_preprocess.cpp:
```diff
@@ -1,8 +1,9 @@
+#include <torch/csrc/jit/backends/backend_debug_handler.h>
 #include <torch/csrc/jit/passes/dead_code_elimination.h>
 #include <torch/csrc/jit/passes/inliner.h>
+#include "torch/csrc/jit/backends/backend_debug_handler.h"
```
delete

raziel replied:
ping

kimishpatel replied:
Sorry. I thought I removed it.

raziel reviewed (changes requested):
Thanks @kimishpatel 

A few small comments, but the main request is to isolate the generation of debug handles to a separate interface and impl.

kimishpatel committed:
Update on "[Pytorch backend delegation] Preprocess to accept"


BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

[ghstack-poisoned]

kimishpatel committed:
Update on "[Pytorch backend delegation] Preprocess to accept"


BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28648613).

raziel commented on torch/csrc/jit/backends/backend_debug_handler.h:
```diff
@@ -114,42 +114,32 @@ using DebugHandleType = int64_t;
 using BackendDebugInfoMapType =
     std::unordered_map<DebugHandleType, DebugInfoTuple>;
 
+using NodeToDebugHandle = std::unordered_map<Node*, DebugHandleType>;
```
I think it'd be cleaner to have these in backend_detail.h

My logic was that since they're mainly for easy reading for the backend developers to have a nice cimple signature for ```BackendPreprocessFunction```

kimishpatel replied:
OK that makes sense.

kimishpatel committed:
Update on "[Pytorch backend delegation] Preprocess to accept"


BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28648613).

raziel reviewed (approved):
Looks great!

@kimishpatel just ping me to review the FB internal changes.

kimishpatel committed:
Update on "[Pytorch backend delegation] Preprocess to accept"


BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

[ghstack-poisoned]

kimishpatel commented:
@kimishpatel has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D28648613).

kimishpatel committed:
Update on "[Pytorch backend delegation] Preprocess to accept"


BackenDebugInforRecorder

Summary:
Prior to this PR:
In order to generate debug handles corresponding to the graph being
lowered, backend's preprocess will call generate_debug_handles and will
get map of Node*-to-debug_handles.
In order to facilitate this, to_backend will own
BackendDebugInfoRecorder and initialize thread local pointer to it.
generate_debug_handle function will query thread local pointer to see if
there is a valid BackendDebugInforRecorder for the context. If there is
it will generate debug handles.

After this PR:
Signature of preprocess is changed such that backends have to register
preprocess that accepts instance of BackendDebugInfoRecorder by
reference. generate_debug_handles is no more a free function but becomes
part of the API of BackendDebugInfoRecorder. Now backend's preprocess
function will call generate_debug_handles on BackendDebugInfoRecorder
instead of free function.

Reason for this change:
With RAII that initializes thread local pointer, results in a lose
contract with backends, which may result in backends not storing
debug information. Making it part of API results in
backends having to be aware of BackendDebugInfoRecorder and explicitly
chosing not to generate/store debug information if they chose to do so.

Test Plan:
backend tests

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D28648613](https://our.internmc.facebook.com/intern/diff/D28648613)

[ghstack-poisoned]
